esmvalcore/_main.py

```diff
@@ -367,102 +365,72 @@ def run(self,
             default (fail if there are any errors),
             strict (fail if there are any warnings).
         """
-        import os
-        import warnings
-
-        from ._config import configure_logging, read_config_user_file
-        from ._recipe import TASKSEP
-        from .cmor.check import CheckLevels
-        from .esgf._logon import logon
-
-        # Check validity of optional command line arguments with experimental
-        # API
-        with warnings.catch_warnings():
-            # ignore experimental API warning
-            warnings.simplefilter("ignore")
-            from .experimental.config._config_object import Config as ExpConfig
-        explicit_optional_kwargs = {
-            'config_file': config_file,
-            'resume_from': resume_from,
-            'max_datasets': max_datasets,
-            'max_years': max_years,
-            'skip_nonexistent': skip_nonexistent,
-            'offline': offline,
-            'diagnostics': diagnostics,
-            'check_level': check_level,
-        }
-        all_optional_kwargs = dict(kwargs)
-        for (key, val) in explicit_optional_kwargs.items():
-            if val is not None:
-                all_optional_kwargs[key] = val
-        ExpConfig(all_optional_kwargs)
+        from .config import CFG
 
         recipe = self._get_recipe(recipe)
-        cfg = read_config_user_file(config_file, recipe.stem, kwargs)
 
+        CFG.load_from_file(config_file)
+        session = CFG.start_session(recipe.stem)
+        if check_level is not None:
+            session['check_level'] = check_level
+        if diagnostics is not None:
+            session['diagnostics'] = diagnostics
+        if max_datasets is not None:
+            session['max_datasets'] = max_datasets
+        if max_years is not None:
+            session['max_years'] = max_years
+        if offline is not None:
+            session['offline'] = offline
+        if skip_nonexistent is not None:
+            session['skip_nonexistent'] = skip_nonexistent
+        session['resume_from'] = parse_resume(resume_from, recipe)
+        session.update(kwargs)
+
+        self._run(recipe, session)
+
+    def _run(self, recipe: Path, session) -> None:
+        """Run `recipe` using `session`."""
         # Create run dir
-        if os.path.exists(cfg['run_dir']):
-            print("ERROR: run_dir {} already exists, aborting to "
-                  "prevent data loss".format(cfg['output_dir']))
-        os.makedirs(cfg['run_dir'])
+        if session.session_dir.exists():
+            print(f"ERROR: output directory {session.session_dir} already"
+                  " exists, aborting to prevent data loss")
```

Comment:
I know that this is the same as before, but I think it would be much clearer to directly raise an error here. Maybe a [`FileExistsError`](https://docs.python.org/3/library/exceptions.html#FileExistsError)?

Reply:
That's exactly the exception that is raised now by the `Path.mkdir()` command below, why do you think it's better to duplicate that code? If we really want to make it nicer, we could raise an error that doesn't show the stack trace because that would be more user friendly, but I would really like to try and limit the number of changes in this pull request because it is already large as it is.

Reply:
It just feels weird that there is a `print` statement that says something about aborting, but then the code continues without an exception or `exit` or anything...It took me some time to realize that the error is implicitly raised in the following line. What's the problem with replacing `print(...)` with `raise FileExistsError(...)`? If you really want to keep it, fine for me.

Reply:
Thanks! I'll keep it as is then, we can always improve later.